File: ts/src/absorb/importers.ts
/**
 * src/absorb/importers.ts
 *
 * Local importers for the absorb command (ADR-055 §4). An importer walks a
 * directory, reads the files relevant to one store kind, and turns each into an
 * AbsorbItem. It NEVER posts anything — the CLI batches the items to the API.
 *
 * Two importers ship in this slice:
 *   claude-code — a Claude Code memory directory: every *.md file, including
 *                 the MEMORY.md index and any CLAUDE.md files.
 *   markdown    — generic catch-all: walk a directory tree for every *.md file.
 *
 * Both reuse the same secret guards (secrets.ts): a credential-named file is
 * skipped, and token-shaped strings are redacted from the content before it is
 * handed back. Each importer reports skipped/redacted counts so the CLI can
 * print an honest summary.
 */

import { readdirSync, readFileSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";

import { redactSecrets, shouldSkipFile } from "./secrets.js";

/** One memory item ready to absorb. Matches the API's AbsorbItem shape. */
export interface AbsorbItem {
  store_kind: string;
  /** Stable identifier within the store — the file's path relative to the root. */
  source_id: string;
  /** First markdown heading, else the filename. */
  title: string;
  /** Redacted file content. */
  content: string;
  /** Filesystem birthtime (ISO-8601), best-effort. */
  created_at?: string;
  /** Filesystem mtime (ISO-8601). */
  updated_at?: string;
}

/** Outcome of walking a directory: items to upload + what was skipped/redacted. */
export interface ScanResult {
  items: AbsorbItem[];
  skippedSecretFiles: string[];
  redactedItems: number;
  redactedSecrets: number;
}

/** Directories never descended into (noise + secret stores). */
const SKIP_DIRS = new Set([
  "node_modules",
  ".git",
  ".venv",
  "venv",
  "__pycache__",
  "dist",
  "build",
  ".next",
  ".cache",
]);

const SUPPORTED_STORES = ["claude-code", "markdown"] as const;
export type StoreKind = (typeof SUPPORTED_STORES)[number];

export function isSupportedStore(kind: string): kind is StoreKind {
  return (SUPPORTED_STORES as readonly string[]).includes(kind);
}

export function supportedStores(): readonly string[] {
  return SUPPORTED_STORES;
}

/** Recursively collect every *.md file under root (relative paths). */
function collectMarkdownFiles(root: string): string[] {
  const out: string[] = [];
  const walk = (dir: string): void => {
    let entries: string[];
    try {
      entries = readdirSync(dir);
    } catch {
      return; // unreadable dir — skip quietly
    }
    for (const entry of entries) {
      const abs = join(dir, entry);
      let st;
      try {
        st = statSync(abs);
      } catch {
        continue;
      }
      if (st.isDirectory()) {
        if (!SKIP_DIRS.has(entry)) walk(abs);
        continue;
      }
      if (st.isFile() && entry.toLowerCase().endsWith(".md")) {
        out.push(relative(root, abs));
      }
    }
  };
  walk(root);
  return out.sort();
}

/** First markdown heading in the content, else the basename without extension. */
function deriveTitle(content: string, relPath: string): string {
  for (const raw of content.split("\n")) {
    const line = raw.trim();
    const m = /^#{1,6}\s+(.+?)\s*#*$/.exec(line);
    if (m && m[1]) return m[1].trim();
  }
  const base = relPath.split(/[\\/]/).pop() ?? relPath;
  return base.replace(/\.md$/i, "");
}

/**
 * Walk a directory and build absorb items for one store kind.
 *
 * Shared by both importers — they differ only in the store_kind tag and (for
 * documentation) the expectation of what lives in the directory; the file
 * selection (every *.md, including MEMORY.md and CLAUDE.md) is identical, which
 * is exactly what a Claude Code memory directory contains.
 */
export function scanDirectory(root: string, storeKind: string): ScanResult {
  const rel = collectMarkdownFiles(root);
  const items: AbsorbItem[] = [];
  const skippedSecretFiles: string[] = [];
  let redactedItems = 0;
  let redactedSecrets = 0;

  for (const relPath of rel) {
    const base = relPath.split(/[\\/]/).pop() ?? relPath;
    if (shouldSkipFile(base)) {
      skippedSecretFiles.push(relPath);
      continue;
    }
    const abs = join(root, relPath);
    let raw: string;
    let st;
    try {
      raw = readFileSync(abs, "utf-8");
      st = statSync(abs);
    } catch {
      continue;
    }
    const { content, redactedCount } = redactSecrets(raw);
    if (redactedCount > 0) {
      redactedItems += 1;
      redactedSecrets += redactedCount;
    }
    // Normalise source_id to forward slashes so it is stable across OSes.
    const sourceId = relPath.split(sep).join("/");
    items.push({
      store_kind: storeKind,
      source_id: sourceId,
      title: deriveTitle(content, relPath),
      content,
      created_at: st.birthtime?.toISOString?.(),
      updated_at: st.mtime?.toISOString?.(),
    });
  }

  return { items, skippedSecretFiles, redactedItems, redactedSecrets };
}

File: py/src/hebbianvault_mcp/absorb/importers.py
"""Directory scanners that turn local Markdown files into absorb items."""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import UTC, datetime
from pathlib import Path
from typing import Literal

from .secrets import redact_secrets, should_skip_file

StoreKind = Literal["claude-code", "markdown"]
SUPPORTED_STORES: tuple[StoreKind, ...] = ("claude-code", "markdown")
SKIP_DIRS = frozenset(
    {"node_modules", ".git", ".venv", "venv", "__pycache__", "dist", "build", ".next", ".cache"}
)


@dataclass(frozen=True)
class AbsorbItem:
    """One redacted local-memory item ready for the absorb API."""

    store_kind: str
    source_id: str
    title: str
    content: str
    created_at: str | None
    updated_at: str

    def as_dict(self) -> dict[str, str]:
        """Return the API payload representation with its snake_case keys."""
        payload = {
            "store_kind": self.store_kind,
            "source_id": self.source_id,
            "title": self.title,
            "content": self.content,
            "updated_at": self.updated_at,
        }
        if self.created_at is not None:
            payload["created_at"] = self.created_at
        return payload


@dataclass(frozen=True)
class ScanResult:
    """Items found and the security actions taken while scanning."""

    items: list[AbsorbItem]
    skipped_secret_files: list[str]
    skipped_symlinks: list[str]
    redacted_items: int
    redacted_secrets: int


def is_supported_store(kind: str) -> bool:
    """Return whether ``kind`` names an available local importer."""
    return kind in SUPPORTED_STORES


def supported_stores() -> tuple[str, ...]:
    """Return the supported store tags in CLI-display order."""
    return SUPPORTED_STORES


def _collect_markdown_files(root: Path) -> tuple[list[Path], list[Path]]:
    """Collect Markdown files below ``root`` and report skipped symlinks."""
    output: list[Path] = []
    skipped_symlinks: list[Path] = []

    def walk(directory: Path) -> None:
        try:
            entries = list(directory.iterdir())
        except OSError:
            return
        for entry in entries:
            try:
                if entry.is_symlink():
                    skipped_symlinks.append(entry.relative_to(root))
                    continue
                if entry.is_dir():
                    if entry.name not in SKIP_DIRS:
                        walk(entry)
                elif entry.is_file() and entry.name.lower().endswith(".md"):
                    output.append(entry.relative_to(root))
            except OSError:
                continue

    walk(root)
    return (
        sorted(output, key=lambda path: path.as_posix()),
        sorted(skipped_symlinks, key=lambda path: path.as_posix()),
    )


def _derive_title(content: str, relative_path: Path) -> str:
    """Use the first Markdown heading, falling back to the filename stem."""
    for raw in content.split("\n"):
        match = re.fullmatch(r"#{1,6}\s+(.+?)\s*#*", raw.strip())
        if match and match.group(1):
            return match.group(1).strip()
    return relative_path.stem


def _iso_timestamp(timestamp: float) -> str:
    """Format a filesystem timestamp as an ISO-8601 UTC value."""
    return datetime.fromtimestamp(timestamp, tz=UTC).isoformat().replace("+00:00", "Z")


def _birthtime_iso(stat: object) -> str | None:
    """Return a filesystem birthtime when the platform exposes one."""
    birthtime = getattr(stat, "st_birthtime", None)
    return _iso_timestamp(birthtime) if birthtime is not None else None


def scan_directory(root: str | Path, store_kind: str) -> ScanResult:
    """Walk a local store and produce redacted items for a supported store kind."""
    root_path = Path(root)
    items: list[AbsorbItem] = []
    skipped_secret_files: list[str] = []
    relative_symlinks: list[Path]
    redacted_items = 0
    redacted_secrets = 0

    relative_paths, relative_symlinks = _collect_markdown_files(root_path)
    for relative_path in relative_paths:
        source_id = relative_path.as_posix()
        if should_skip_file(relative_path.name):
            skipped_secret_files.append(source_id)
            continue
        try:
            path = root_path / relative_path
            raw = path.read_text(encoding="utf-8")
            stat = path.stat()
        except OSError:
            continue
        redaction = redact_secrets(raw)
        if redaction.redacted_count:
            redacted_items += 1
            redacted_secrets += redaction.redacted_count
        items.append(
            AbsorbItem(
                store_kind=store_kind,
                source_id=source_id,
                title=_derive_title(redaction.content, relative_path),
                content=redaction.content,
                created_at=_birthtime_iso(stat),
                updated_at=_iso_timestamp(stat.st_mtime),
            )
        )

    return ScanResult(
        items=items,
        skipped_secret_files=skipped_secret_files,
        skipped_symlinks=[path.as_posix() for path in relative_symlinks],
        redacted_items=redacted_items,
        redacted_secrets=redacted_secrets,
    )
